File: cpp/leetcode/editor/cn/[83]删除排序链表中的重复元素.cpp
#include <vector>
#include "util.h"
#include <map>
#include <string>
#include <utility>
#include <memory>

using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode *deleteDuplicates(ListNode *head) {
        if (head == nullptr) return nullptr;

        auto slow = head;
        auto fast = head;
        while (fast != nullptr) {
            if (slow->val != fast->val) {
                slow->next = fast;
                slow = slow->next;
            }
            fast = fast->next;
        }
        slow->next = nullptr;
        return head;
    }
};
//leetcode submit region end(Prohibit modification and deletion)


int main(int argv, char *argc[]) {
}

File: scala/leetcode_scala/src/main/scala/leetcode/editor/cn/[83]删除排序链表中的重复元素.scala
package leetcode.editor.cn
object RemoveDuplicatesFromSortedList {
//leetcode submit region begin(Prohibit modification and deletion)
  /** Definition for singly-linked list.
    * class ListNode(_x: Int = 0, _next: ListNode = null) {
    *   var next: ListNode = _next
    *   var x: Int = _x
    * }
    */
  object Solution {
    def deleteDuplicates(head: ListNode): ListNode = {

      @annotation.tailrec
      def impl(slow: ListNode, fast: ListNode): ListNode = {
        if (fast == null) slow
        else {
          if (slow.x == fast.x) {
            impl(slow, fast.next)
          } else {
            slow.next = fast
            impl(fast, fast.next)
          }
        }
      }

      if(head == null) null
      else {
        val r = impl(head, head)
        r.next = null
        head
      }
    }
  }
//leetcode submit region end(Prohibit modification and deletion)

}
